pkg/core/validators/types.go

```diff
@@ -1,14 +1,64 @@
 package validators
 
 import (
-	"github.com/pkg/errors"
-	"strings"
+	"fmt"
 )
 
-func NewValidationError(err error) error {
-	return errors.Wrap(err, "validation error")
+type ValidationError struct {
+	Violations []Violation
+}
+
+type Violation struct {
+	Field   string
+	Message string
+}
+
+func (v *ValidationError) Error() string {
+	err := ""
+	for _, violation := range v.Violations {
+		if err != "" {
+			err = fmt.Sprintf("%s; %s: %s", err, violation.Field, violation.Message)
+		} else {
+			err += fmt.Sprintf("%s: %s", violation.Field, violation.Message)
+		}
+	}
+	return err
+}
+
+func (v *ValidationError) HasViolations() bool {
+	return len(v.Violations) > 0
+}
+
+func (v *ValidationError) ToError() error {
```

Comment:
Method name is confusing (`ValidationError` is already an `error`).

Maybe `EmptyToNil()` or `OrNil()` ?